library/std/src/panicking.rs

```diff
@@ -261,7 +262,11 @@ fn default_hook(info: &PanicInfo<'_>) {
 
         let mut print_snippet = || -> Result<(), crate::io::Error> {
             let filename = location.file();
-            if let Ok(file) = File::open(filename) {
+            let path = PathBuf::from(filename);
+            // We check if `path` is a symlink to avoid people creating a symlink for a file they
+            // don't have permissions, but that the running binary does, to use it as an
+            // exfiltration vector.
+            if !path.is_symlink() && let Ok(file) = File::open(path) {
```

Comment:
This doesn't help if one of the directories in the path is a symlink. You need O_NOFOLLOW to handle directories without a TOCTOU bug.

Reply:
Changed. We'd need something equivalent for Windows as well.

Reply:
Ended up pulling back from this and instead walked up the path and checked every one for being a symlink with `is_symlink` because that will work in all platforms.

Reply:
This returned the TOCTOU bug though, didn't it? The change may happen between the check and `File::open` call.

Reply:
That is true. The [closest to a reasonable fallback](https://www.gnu.org/software/findutils/manual/html_node/find_html/Systems-without-O_005fNOFOLLOW.html) seems to be equivalent to "check if_symlink, open file, check again", which is still susceptible to timing attacks, just less so. I can do that (and try to use `O_NOFOLLOW` on platforms that support it).